node/src/p2p.rs

```diff
@@ -547,17 +576,45 @@ where
     }
 
     #[instrument(skip_all)]
-    fn on_header_sub_message(&mut self, data: &[u8]) {
+    async fn on_header_sub_message(&mut self, data: &[u8]) -> gossipsub::MessageAcceptance {
         let Ok(header) = ExtendedHeader::decode_and_validate(data) else {
             trace!("Malformed or invalid header from header-sub");
-            return;
+            return gossipsub::MessageAcceptance::Reject;
         };
 
-        debug!("New header from header-sub ({header})");
-        // TODO: inform syncer about it
-        //
-        // TODO: Verify header against store. If it's verified propagate
-        // it to other peers. If not, discard it.
+        trace!("Received header from header-sub ({header})");
+
+        // We care only for headers newer than head
+        let Ok(head) = self.store.get_head().await else {
+            trace!("Store not initialized yet. Ignore header from header-sub ({header}).");
+            return gossipsub::MessageAcceptance::Ignore;
+        };
+
+        // Verify header
+        if head.verify(&header).is_err() {
+            trace!("Verification failed. Ignore header from header-sub ({header}).");
+            return gossipsub::MessageAcceptance::Ignore;
+        }
+
+        // Update it only if needed
+        let updated = self.header_sub_watcher.send_if_modified(|state| {
+            if let Some(known_header) = state {
+                if known_header.height() >= header.height() {
+                    trace!("Header-sub already up-to-date. Ignore header ({header}).");
+                    return false;
+                }
+            }
+
+            debug!("New header from header-sub ({header})");
+            *state = Some(header);
+            true
+        });
+
+        if updated {
+            gossipsub::MessageAcceptance::Accept
+        } else {
+            gossipsub::MessageAcceptance::Ignore
```

Comment:
I think after the `verify` succeeds we should `Accept` unconditionally
https://github.com/celestiaorg/go-header/blob/main/p2p/subscriber.go#L84

Reply:
Wouldn't this spam the network?

Reply:
Hmm. If you notice the `val` is a callback function and ends up [here](https://github.com/celestiaorg/go-header/blob/e75d00df4c66d32f8528185d9e91273d10db94b5/sync/sync_head.go#L137). Then they [verify it with the current subjective head](https://github.com/celestiaorg/go-header/blob/e75d00df4c66d32f8528185d9e91273d10db94b5/sync/sync_head.go#L168) instead of the `Store` head. Which means older heights are ignored.

I can change the verification to be done on the current subjective head.

Reply:
hmm true. yeah, I think we should verify with syncer's subjective head when syncer is introduced